File: backend/app/api/schemas/projects.py
"""API schemas for uploaded audit projects."""
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel

from app.domain.projects import ProjectEvent, ProjectRecord, ProjectStatus


class ProjectEventResponse(BaseModel):
    event_id: int
    stage: str
    message: str
    completed_steps: int
    total_steps: int
    warning: bool
    occurred_at: datetime

    @classmethod
    def from_domain(cls, event: ProjectEvent) -> "ProjectEventResponse":
        return cls(
            event_id=event.event_id,
            stage=event.stage,
            message=event.message,
            completed_steps=event.completed_steps,
            total_steps=event.total_steps,
            warning=event.warning,
            occurred_at=event.occurred_at,
        )


class ProjectResponse(BaseModel):
    project_id: str
    name: str
    source_type: str
    status: ProjectStatus
    current_activity: str | None
    allowed_actions: list[str]
    created_at: datetime
    updated_at: datetime
    started_at: datetime | None
    completed_at: datetime | None
    output_available: bool
    output_download_url: str | None
    version: str | None
    issue_count: int | None
    error: str | None
    raw_expires_at: datetime | None
    raw_deleted_at: datetime | None

    @classmethod
    def from_domain(cls, project: ProjectRecord) -> "ProjectResponse":
        output_available = (
            project.status == ProjectStatus.COMPLETED
            and project.output_path is not None
        )
        return cls(
            project_id=project.project_id,
            name=project.name,
            source_type=project.source_type,
            status=project.status,
            current_activity=project.current_activity,
            allowed_actions=_allowed_actions(project),
            created_at=project.created_at,
            updated_at=project.updated_at,
            started_at=project.started_at,
            completed_at=project.completed_at,
            output_available=output_available,
            output_download_url=(
                f"/api/v1/projects/{project.project_id}/output"
                if output_available
                else None
            ),
            version=project.version,
            issue_count=project.issue_count,
            error=project.error,
            raw_expires_at=project.raw_expires_at,
            raw_deleted_at=project.raw_deleted_at,
        )


def _allowed_actions(project: ProjectRecord) -> list[str]:
    actions = ["VIEW_STATUS"]
    if project.status == ProjectStatus.PROCESSING:
        actions.append("VIEW_PROGRESS")
    elif project.status == ProjectStatus.COMPLETED:
        actions.extend(["VIEW_PROGRESS", "DOWNLOAD_OUTPUT"])
    elif project.status == ProjectStatus.FAILED:
        actions.append("VIEW_PROGRESS")
    return actions

File: frontend/src/shared/types/projects.ts
export type ProjectStatus =
  | "UPLOADING"
  | "PROCESSING"
  | "COMPLETED"
  | "FAILED"
  | "READY_FOR_DISCOVERY"
  | "CANDIDATES_AVAILABLE"
  | "OUTPUT_AVAILABLE"

export type ProjectAction = "VIEW_STATUS" | "VIEW_PROGRESS" | "DOWNLOAD_OUTPUT"

export interface AuditProject {
  project_id: string
  name: string
  source_type: string
  status: ProjectStatus
  current_activity: string | null
  allowed_actions: ProjectAction[]
  created_at: string
  updated_at: string
  started_at: string | null
  completed_at: string | null
  output_available: boolean
  output_download_url: string | null
  version: string | null
  issue_count: number | null
  error: string | null
  raw_expires_at: string | null
  raw_deleted_at: string | null
}

export interface ProjectEvent {
  event_id: number
  stage: string
  message: string
  completed_steps: number
  total_steps: number
  warning: boolean
  occurred_at: string
}

export interface ApiErrorBody {
  error?: {
    code?: string
    message?: string
    details?: Record<string, unknown>
    correlation_id?: string | null
  }
}

export interface UploadProjectInput {
  name: string
  files: File[]
  auditorIssues: AuditorIssueInput[]
}

export interface AuditorIssueInput {
  title_hint: string
  observed_gap: string
  evidence_summary: string
  evidence_refs: string[]
  sop_refs: string[]
  risk_category: string
}

export type UploadSessionState =
  | "UPLOADING"
  | "VALIDATING"
  | "READY_TO_CREATE"
  | "INVALID"
  | "PROMOTED"
  | "EXPIRED"

export type LogicalRole = "SCOPE" | "RISK_CONTEXT" | "EVIDENCE" | "CRITERIA" | "CONTEXT"

export interface SourceFile {
  document_id: string
  name: string
  relative_path: string
  logical_role: LogicalRole
  size_bytes: number
  content_type: string | null
  status: string
  parse_status: string
}

export interface SourceFolder {
  name: string
  logical_role: LogicalRole
  file_count: number
  files: SourceFile[]
}

export interface SourceTree {
  snapshot_id: string
  status: string
  folder_count: number
  file_count: number
  total_size_bytes: number
  folders: SourceFolder[]
}

export interface UploadSessionFile {
  file_id: string
  relative_path: string
  size_bytes: number
  content_type: string | null
  modified_at: string | null
  upload_status: string
  logical_role: LogicalRole | null
  readability_status: string | null
  validation_message: string | null
  upload_method: string
  upload_url: string
  required_headers: Record<string, string>
}

export interface UploadSession {
  session_id: string
  state: UploadSessionState
  created_at: string
  expires_at: string
  promoted_at: string | null
  files: UploadSessionFile[]
  validation_report: UploadValidationReport | null
  allowed_actions: string[]
  action_reasons: Record<string, string>
}

export interface UploadValidationMessage {
  code: string
  message: string
  file_id: string | null
  relative_path: string | null
  blocking: boolean
  details?: Record<string, unknown>
}

export interface UploadValidationReport {
  valid: boolean
  errors: UploadValidationMessage[]
  warnings: UploadValidationMessage[]
  role_summary: Record<Exclude<LogicalRole, "CONTEXT">, number>
}

export type AuditVersionState =
  | "DRAFT"
  | "CANDIDATES_READY"
  | "AUDITING"
  | "DOCX_READY"
  | "STALE_OUTPUT"

export type JobState = "QUEUED" | "RUNNING" | "SUCCEEDED" | "INCOMPLETE" | "FAILED"
export type JobType = "DISCOVERY" | "AUDIT"

export interface AuditJob {
  job_id: string
  project_id: string
  project_version_id: string
  job_type: JobType
  state: JobState
  stage: string | null
  completed_items: number
  total_items: number | null
  current_message: string | null
  attempt_count: number
  correlation_id: string
  created_at: string
  updated_at: string
  heartbeat_at: string | null
  error: string | null
}

export interface ProjectVersion {
  version_id: string
  project_id: string
  sequence_no: number
  label: string
  base_version_id: string | null
  state: AuditVersionState
  issue_revision: number
  issue_counts: Record<string, number>
  latest_job: AuditJob | null
  output_available: boolean
  allowed_actions: string[]
  created_at: string
  updated_at: string
}

export type IssueOrigin = "AI_DISCOVERED" | "MANUAL"
export type IssueStatus =
  | "DRAFT"
  | "READY_FOR_REVIEW"
  | "APPROVED"
  | "NEEDS_EVIDENCE"
  | "REJECTED"
  | "OUT_OF_SCOPE"

export interface SourceReference {
  reference_id: string
  ref_kind: "EVIDENCE" | "CRITERIA"
  document_id: string
  unit_id: string | null
  location: Record<string, unknown>
  quote: string | null
}

export interface CandidateIssue {
  issue_id: string
  project_version_id: string
  origin: IssueOrigin
  status: IssueStatus
  observed_gap: string
  title_hint: string | null
  evidence_summary: string | null
  risk_category: string | null
  confidence: number | null
  validation_flags: string[]
  row_version: number
  source_refs: SourceReference[]
  created_at: string
  updated_at: string
}

export interface OutputRevision {
  output_id: string
  project_version_id: string
  ordinal: number
  status: "CURRENT" | "STALE"
  filename: string
  content_hash: string
  created_at: string
  download_url: string
}

export interface CreatedAuditProject {
  project_id: string
  name: string
  state: "READY_FOR_DISCOVERY" | "CANDIDATES_AVAILABLE" | "OUTPUT_AVAILABLE"
  source_snapshot_id: string
  version: {
    version_id: string
    sequence_no: number
    label: string
    state: AuditVersionState
    issue_revision: number
  }
  created_at: string
  updated_at: string
}
